infrastructures/docker/environment/src/main/java/org/eclipse/che/workspace/infrastructure/docker/environment/compose/ComposeEnvironmentFactory.java

```diff
@@ -116,14 +121,25 @@ void addRamLimitAttribute(
         machineConfig = new InternalMachineConfig();
         machines.put(entry.getKey(), machineConfig);
       }
+      Long memLimit = entry.getValue().getMemLimit();
+      Long memRequest = entry.getValue().getMemRequest();
+      // if both properties are not defined
+      if ((memLimit == null || memLimit <= 0) && (memRequest == null || memRequest <= 0)) {
```

Comment:
Should be the same be done for other environment factories?
Maybe it can be used in a separated component like
```java
class MemoryAttributesProvisioner {
  ...
   MemoryAttributesProvisioner(
      @Named("che.workspace.default_memory_limit_mb") long defaultMachineMaxMemorySizeMB,
      @Named("che.workspace.default_memory_request_mb") long defaultMachineRequestMemorySizeMB) {
    ....
    }

    void provision(MachineConfig machine, @Nullable Long memoryLimit, @Nullable Long memoryRequest) { // where memLimit and memRequest is provided by infrastructure and fetched from recipe
       // evaluate here memory settings configured or default and set correct one to machine attributes if needed
    }
}
```
@garagatyi WDYT?

Reply:
you are talking about DockerImageEnvironmentFactory and DockerfileEnvironmentFactory right ?
the logic there currently only sets values if they are not defined in machineConfig.
so the values in the machine config at that point are the values requested by the user right ?
and they might not be valid correct ?

Reply:
> you are talking about DockerImageEnvironmentFactory and DockerfileEnvironmentFactory right ?

Yes, and KubernetesEnvironmentFactory and DockerfileEnvironmentFactory as well.

> the logic there currently only sets values if they are not defined in machineConfig.

Well, actually there can be a conflict situation which I described at the beginning. What if one attribute is configured and another one is missing, then default one should be used and conflict may happen. I hope the detailed explanation you can find here https://github.com/eclipse/che/pull/10673#discussion_r208473063.

> so the values in the machine config at that point are the values requested by the user right ?

right

> and they might not be valid correct?

Well, I missed this point, but it's a good one =)
It makes sense to add validation into `MachineConfigsValidator` to ensure that at that point provided values are correct.

Reply:
I see that on k8sEnvironmentFactory and OSEnvFactory you guys use InternalMachineConfig and it does not implement MachineConfig interface. Is there a reason for it ? 
@sleshchenko  should i pass InternalMachineConfig to the provision method in your suggestion above then ?